### PR TITLE
Additional GAGS configuration json verification

### PR DESCRIPTION
**Pull request by ninjanomnom**
This may be a bit overkill.

This standardizes how values are read from the json for greyscale layers so that error handling can check for some additional things:
No extra keys in the json that are unknown
No missing keys that a layer needs to work
Values are the expected type for that key

A variety of error messages have been added for various ways the json can be malformed and should hopefully provide good feedback for anyone working with greyscale configurations.

No player facing changes.

### CODE REVIEW

**Comment by coiax** [comment]
At the risk of asking a stupid question, is it really important that we do the validation of GAGs ingame and with DM?

Feel like something like `jsonschema` could verify this in linting, and a lot easier.

**Comment by ninjanomnom** [comment]
To summarize a discussion on discord:

While we could have a separate linting step for json schema there are a few issues with being able to replace this with it. 

The first is that part of the point of the error messages is to help in development of gags sprites, especially for spriters. Running the lints though is not something most of them know how to do or check for.

Secondly, an external system verifying the json won't be connected to how layers are actually read and can easily get out of date with the code. The way this is set up the way the json is read and the verification will always be in sync. 

Finally, the verification step also converts values and verifies using runtime data such as the reference layer checking that the referenced gags config is valid and set up.

File: code/datums/greyscale/json_reader.dm [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/// Takes a json list and extracts a single value.
+/// Subtypes represent different conversions of that value.
+/datum/json_reader
+
+/datum/json_reader/proc/ReadJson(value)
```
Comment (Mothblocks) [inline code/datums/greyscale/json_reader.dm:5]: Document as subtypes use it, also please add in a `return`, it doesn't do anything but it just looks more correct to me 😛

File: code/datums/greyscale/layer.dm [hunk 2]
```diff
@@ -3,26 +3,52 @@
...
+		for(var/path in subtypesof(/datum/json_reader))
+			json_readers[path] = new path
+
+	json_data -= "type" // This is used to look us up and doesn't need to be verified like the rest of the data
+
+	ReadJsonData(json_data)
+	Initialize(icon_file)
+
+/// Override this to do initial set up
+/datum/greyscale_layer/proc/Initialize(icon_file)
```
Comment (Mothblocks) [inline code/datums/greyscale/layer.dm:20]:
```suggestion
/datum/greyscale_layer/proc/Initialize(icon_file)
	return
```

Same here, just looks more correct.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Cool stuff, pedantic notes.

**Pull request merged**